scripts/migrations/2020-10-miro-migration/transfer_packager.py

```diff
@@ -0,0 +1,195 @@
+                for task in itertools.islice(s3_key_list_iter, S3_DOWNLOAD_CONCURRENCY)
+            }
+
+            while futures:
+                # Wait for the next future to complete.
+                done, _ = concurrent.futures.wait(
+                    futures, return_when=concurrent.futures.FIRST_COMPLETED
+                )
+
+                for fut in done:
+                    futures.pop(fut)
+
+                pbar.update(len(done))
+
+                # Schedule the next set of futures.  We don't want more than N futures
+                # in the pool at a time, to keep memory consumption down.
+                for task in itertools.islice(s3_key_list_iter, len(done)):
+                    fut = executor.submit(_get_s3_object, task)
+                    futures[fut] = task
+
+        target_folder_count = sum(
+            [len(files) for r, d, files in os.walk(target_folder)]
+        )
+
+        assert target_folder_count == len(s3_key_list), (
+            f"Unexpected file count in {target_folder}: "
+            f"{target_folder_count} != {len(s3_key_list)}"
+        )
+
+
+def _compress_folder(target_folder, remove_folder=True):
+    archive_name = shutil.make_archive(target_folder, "zip", target_folder)
+    if remove_folder:
+        shutil.rmtree(target_folder, ignore_errors=True)
+
+    return archive_name
+
+
+def _create_metadata(target_folder, group_name):
+    metadata_folder = os.path.join(target_folder, "metadata")
+    metadata_file_location = os.path.join(metadata_folder, "metadata.csv")
+
+    os.makedirs(metadata_folder, exist_ok=True)
+
+    metadata_lines = ["filename,dc.identifier,isMiro", f"objects/,{group_name},true"]
+
+    metadata_contents = "\n".join(metadata_lines)
+
+    with open(metadata_file_location, "w") as fp:
+        fp.write(metadata_contents)
+
+
+def upload_transfer_package(
+    s3_client, s3_bucket, s3_path, transfer_package, cleanup=False
+):
+    file_location = transfer_package.local_location
+
+    filename = os.path.basename(file_location)
+    s3_key = f"{s3_path}/{filename}"
+
+    bytes_to_upload = os.path.getsize(file_location)
+
+    with tqdm(total=bytes_to_upload) as pbar:
+
+        def _update_pbar(num_bytes):
+            pbar.update(num_bytes)
+
+        s3_client.upload_file(
+            Filename=file_location, Bucket=s3_bucket, Key=s3_key, Callback=_update_pbar
+        )
+
+        s3_content_length = get_s3_content_length(
+            s3_client=s3_client, s3_bucket=s3_bucket, s3_key=s3_key
+        )
+
+        _check_local_content_length(
+            file_location=file_location, expected_content_length=s3_content_length
+        )
+
+        if cleanup:
+            os.remove(file_location)
+
+    transfer_package.s3_location = s3_key
+
+    return transfer_package
+
+
+def create_transfer_package(s3_client, group_name, s3_bucket, s3_key_list, prefix):
+    target_base_folder = "target"
+    folder_id = str(uuid.uuid4())
```

Comment:
I would throw at https://leancrew.com/all-this/2014/10/asciifying/ at the group name